libr/bin/p/bin_p9.c

```diff
@@ -247,66 +272,217 @@ static RList *sections(RBinFile *bf) {
-	RList *ret = NULL;
-	RBinPlan9Obj *o = (RBinPlan9Obj *)bf->o->bin_obj;
+typedef struct {
+	ut64 value;
+	char type;
+	char *name;
+} Sym;
 
-	if (!(ret = r_list_newf (free))) {
-		return NULL;
+static st64 sym_read(RBinFile *bf, Sym *sym, const ut64 offset) {
+	st64 size = 0;
+	const RBinPlan9Obj *o = (RBinPlan9Obj *)bf->o->bin_obj;
+	const ut64 syms = o->header_size + o->header.text + o->header.data;
+
+	ut64 value;
+	if (o->header.magic & HDR_MAGIC) {
+		// for 64-bit binaries the value type is 8 bytes
+		value = r_buf_read_be64_at (bf->buf, syms + offset);
+		if (value == UT64_MAX) {
+			return -1;
+		}
+		size += sizeof (ut64);
+	} else {
+		value = (ut64)r_buf_read_be32_at (bf->buf, syms + offset);
+		if (value == UT32_MAX) {
+			return -1;
+		}
+		size += sizeof (ut32);
 	}
 
-	ut64 syms = o->header_size + o->header.text + o->header.data;
+	const ut8 typ = r_buf_read8_at (bf->buf, syms + offset + size);
+	if (typ == UT8_MAX) {
+		return -1;
+	}
+	size += sizeof (typ);
+	const char type = typ & 0x7f;
 
-	ut64 offset = 0;
-	while (offset < o->header.syms) {
-		ut64 value;
-		if (o->header.magic & HDR_MAGIC) {
-			// for 64-bit binaries the value type is 8 bytes
-			value = r_buf_read_be64_at (bf->buf, syms + offset);
-			if (value == UT64_MAX) {
-				goto error;
+	char *name = r_buf_get_string (bf->buf, syms + offset + size);
+	if (!name) {
+		return -1;
+	}
+	size += strlen (name) + 1;
+
+	sym->value = value;
+	sym->type = type;
+	sym->name = name;
+	return size;
+}
+
+static void sym_fini(void *sym, void *user) {
+	(void)user;
+	Sym *s = (Sym *)sym;
+	if (s && s->name) {
+		free (s->name);
+		s->name = NULL;
+	}
+}
+
+static int apply_history(RBinFile *bf, ut64 pc, ut64 line, Sym *base, Sym **ret) {
+	// start of current level
+	Sym *start;
+	// current entry
+	Sym *h;
+	// sum of size of files this level
+	st64 delta;
+	int k;
+
+	// see fline in libmach
+	start = base;
+	h = base;
+	delta = h->value;
+	while (h && h->name && line > h->value) {
+		if (strlen (h->name)) {
+			if (start == base)
+				start = h++;
+			else {
+				k = apply_history (bf, pc, line, start, &h);
+				if (k <= 0) {
+					return k;
+				}
 			}
-			offset += sizeof (ut64);
 		} else {
-			value = (ut64)r_buf_read_be32_at (bf->buf, syms + offset);
-			if (value == UT32_MAX) {
-				goto error;
+			if (start == base) {
+				// end of recursion level
+				if (ret) {
+					*ret = h;
+				}
+				return 1;
+			} else {
+				// end of included file
+				delta += h->value - start->value;
+				h++;
+				start = base;
 			}
-			offset += sizeof (ut32);
 		}
+	}
+
+	if (!h) {
+		return -1;
+	}
+
+	char *name = h->name? start->name: "<unknown>";
+	if (start != base) {
+		line = line - start->value + 1;
+	} else {
+		line = line - delta + 1;
+	}
+
+	char buffer[SDB_NUM_BUFSZ] = {0};
+	char *line_info = r_str_newf ("%s|%"PFMT64d, name, line);
+	char *address = sdb_itoa (pc, 16, buffer, sizeof (buffer));
+	sdb_set (bf->sdb_addrinfo, address, line_info, 0);
```

Comment:
i did new sdb apis where the key is a number, you can use `sdb_nset` and remove the sdb_itoa lines